solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrJdkClient.java

```diff
@@ -287,6 +278,64 @@ private HttpResponse<InputStream> doPutOrPost(
     return response;
   }
 
+  private volatile boolean headRequested;
+  private volatile boolean headSucceeded;
```

Comment:
volatile isn't needed because it's only accessed by one method (which ought to be mentioned in a comment so this invariant holds)

Reply:
I think these booleans need to go away anyhow.  My reasoning here is we cannot guarantee the client will behave the same every time with this 1-shot try with a HEAD request.  What if there is a network issue?  Then that client instance is stuck on Http/1.1 for its lifetime for these requests.

I think it is probably best to issue the HEAD request every time, to guarantee consistency.  This only affects POST requests with a chunked body (updates, not queries).  It only will be needed if not TLS, and if the user does not force Http/1.1.  I do not think it adds anymore overhead than if it always issued an upgrade request with no body (as the Eclipse maintainers seem to think it ought to.)  We can put a note about this in the Javadoc and if none of these options are workable for any user, they always have `Http2SolrClient` to fall back on.  

Progress, not Perfection, eh?

Reply:
Does a "HEAD request every time" -- do you mean when the client is opened/instantiated?  Fine; agree to "progress not perfection".  It can be improved later driven by actual vs perceived problems.

Reply:
I'm ok with HEAD every time as a start. Then perhaps we can make it more clever over time. And if/when JDK improves, we can detect JDK version and conditionally force HTTP/2.

This also means we can avoid sending ping in constructor, right?

Reply:
By "every time" I mean with every POST request that meets the criteria.  (chunked body; no TLS; not forcing Http/1.1).  This generally will mean every Solr Update will incur both an Http HEAD and an Http POST.  This should be no different if the JDK Client behaved as the Jetty maintainers think it should:  to incur both an Empty Upgrade then the Actual Request.  It is true most of the time the Http HEAD request is unnecessary.  However, I do not know how we know _when_ it is unnecessary.  So at the expense of performance, we get it to work correctly on a consistent basis.  

Maybe for a Version Two, we could get into retry logic, where if it fails, then try a HEAD request and then try the POST once again.  But the more I think about what it would take to do this right the less motivated I feel to go down this path.

Reply:
I'd much prefer the boolean return to ensure we do this once and not every time.  Any issues with that?

Reply:
The issue with doing it just once is it cannot guarantee the client will behave the same every time.  If we try once with the first request, we are at the mercy of luck as to whether or not it could fail for some unrelated reason.  Then the client is downgraded to Http/1.1 for its lifetime (which could be the life of the application).  I can see future bug reports along the lines of "sometimes, after application restart, my solr client downgrades to Http/1.1 for an unknown reason and then I have to restart again to fix it."

I think the main use-case for chunked body POST requests are index updates, and most users would batch these, so the cost of an additional HEAD request might be negligible?  

Maybe I am missing something here, perhaps there is a better way to ensure consistency without adding (much) cost?

Reply:
I don't think anybody would notice an HTTP 2 to HTTP 1.1 fallback.  It works.

Reply:
I left this as it is and it only will complete a HEAD request the very first time a POST request with a chunked body is executed.  If the HEAD request succeeds it lets the jdk client send upgrade requests from then on.  If it fails, it forces http/1.1 for all POST requests with a chunked body.  I did add an INFO logging message so users might get a clue of what is going on in these instances.

None of this applies if the request uses TLS.  For secure requests, Jetty and the JDK are in agreement on the correct way to negotiate the protocol so all of this is unnecessary.  Of course, if the user requests to always downgrade to http/1.1 this log is skipped as well.